BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,180 @@
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOC_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+    }
+
+    // i2c config
+    i2c_struct.Init.ClockSpeed = I2C_CLOCK_SPEED;
+    i2c_struct.Init.DutyCycle = I2C_DUTYCYCLE_2;
+    // 7 bit addr mode
+    i2c_struct.Init.OwnAddress1 = 0x50 << 1;
+    i2c_struct.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    i2c_struct.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    i2c_struct.Init.GeneralCallMode = (1 << 6); 
+    i2c_struct.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+
+    // gpio config
+    gpio_struct.Pin = GPIO_PIN_8; 
+    gpio_struct.Mode = MODE_AF;
+    gpio_struct.Pull = GPIO_PULLUP;
+    gpio_struct.Speed = GPIO_SPEED_MEDIUM; 
+    gpio_struct.Alternate = GPIO_AF4_I2C3;
+
+    // PA8 is SCL
+    HAL_GPIO_Init(GPIOA, &gpio_struct);
+
+    // PC9 is SDA
+    gpio_struct.Pin = GPIO_PIN_9;
+    HAL_GPIO_Init(GPIOC, &gpio_struct);
+}
+
+/** 
+* @param    hi2c:            to structure that has config data
+* @param    deviceAdd :      target device address
+* @param    pDataBuff :      data buffer
+* @param    len :           amount of data   
+* @return   HAL Status    
+*/
+void I2CWrite(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAdd, 
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAdd << 1, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) > len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaQueue, deviceAdd << 1, 0);
+            xQueueSend(I2C_MetaQueue, len, 0);
+    }
+}
+
+//overwrite this
+void HAL_I2C_MasterTxCpltCallback(I2C_HandleTypeDef *hi2c) {
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+    if (xQueueIsQueueEmptyFromISR(I2C_MetaQueue) == pdFALSE) {
+        uint32_t deviceAdd, len;
+        uint8_t temp;
+        xQueueReceiveFromISR(I2C_MetaQueue, &deviceAdd, &temp);
+        xQueueReceiveFromISR(I2C_MetaQueue, &len, &temp);
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueReceiveFromISR(I2C_DataQueue, &(buffer[i]), &temp);
+        }
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAdd, buffer, len);
+    }
+}
+
+/** 
+* @param    hi2c:            to structure that has config data
+* @param    deviceAdd :      target device address
+* @param    memoryAdd :      
+* @param    memoryAddSize :  
+* @param    pDataBuff :      data buffer
+* @param    len :            amount of data   
+* @return   HAL Status    
+*/
+void I2CRegisterWrite(I2C_HandleTypeDef *hi2c, 
```

Comment:
These functions should probably be renamed to BSP_I2C_XXXXX to follow convention. See the comment on the header file for more info

Reply:
Also make sure the comment made for each function has descriptions for each parameter